ufl/algorithms/apply_derivatives.py

```diff
@@ -84,6 +86,17 @@
 # - ReferenceDivRuleset
 
 
+def flatten_domain_element(domain, element):
+    """Return the flattened (domain, element) pairs for mixed domain problems."""
+    if not isinstance(domain, MixedMesh):
+        return ((domain, element),)
+    flattened = ()
+    assert len(domain) == len(element.sub_elements)
+    for d, e in zip(domain, element.sub_elements):
+        flattened += flatten_domain_element(d, e)
+    return flattened
```

Comment:
Typing + specification of what is returned if the problem is not a "mixed domain problem"/